services/QuillLMS/client/app/bundles/Grammar/actions/session.ts

```diff
@@ -166,27 +166,35 @@ export const startListeningToFollowUpQuestionsForProofreaderSession = (proofread
   }
 }
 
-const handleProofreaderSession = (proofreaderSession, state) => {
+export const handleProofreaderSession = (proofreaderSession, state) => {
   return dispatch => {
 
     // if there is no proofreader session or concept results for that session, we can't do anything here
     // if there are already answered questions or the session is already in the state
     // that all gets processed by the normal `setSessionReducerToSavedSession` function and we don't need to set new ones
     if (!proofreaderSession || !proofreaderSession.conceptResults || proofreaderSession.answeredQuestions || state.session.proofreaderSession) { return }
 
-    const concepts: { [key: string]: { quantity: 1|2|3 } } = {}
-    const incorrectConcepts = proofreaderSession.conceptResults.filter(cr => cr.metadata.correct === 0)
+    const concepts: { [key: string]: any } = {}
```

Comment:
I found the changing of type for `concepts` value to be confusing.  I would introduce an intermediate variable to hold the incorrectConceptUIDs:

```
const incorrectConceptUIDs = proofreaderSession.conceptResults.reduce((results, cr) => {
  if (metadata.correct === 0 && !results.includes(cr.concept_uid)) {  results.push(cr.concept_uid); }
  return results;
}, []);


const quantityMap = { '1-4': 3, '5-9': 2, '10+': 1 };  // These magic might be worth some explanation:
const quantity = incorrectConceptUIDs.length > 9 ? quantityMap['10+'] 
               : incorrectConceptUIDs.length > 4 ? quantityMap['5-9']
               : quantityMap['1-4']
             
 const concepts: { [key: string]: { quantity: 1|2|3 } } = incorrectConceptUIDs.reduce((results, cr) => {
    if (!results[cr.concept_uid]) { concepts[cr.concept_uid] = { quantity }; }
    return concepts;
  }, {});              
 ```

Reply:
I like this idea-- I do just want to note that concepts results can have the same concept so that's why I opted for a hash. I could just put `incorrectConceptUIDs` inside of a Set though to remove duplicates

Reply:
I like this idea-- I'll add a comment for some further explanation

Reply:
I might be mixing something up here, but I think `incorrectConceptUIDs` should not have duplicates due to `!results.includes(cr.concept_uid)`